### ISSUE TITLE
Plain SQL templates improperly recognise "?<>" as a PostgreSQL operator

### ISSUE THREAD

**Issue by kochetkovSergeY** [jOOQ/jOOQ#11442]
### Expected behavior
Query executes successfully

### Actual behavior
Query fails with error that there is no parameter value fo second parameter

> Caused by: org.postgresql.util.PSQLException: No value specified for parameter 2.

### Steps to reproduce the problem
Sql query string:

```java
String sql="SELECT CASE\n"+ "WHEN  ? ='REASON1'  and ?<> 'FAILURE'  THEN 'STATUS2'\n"
                                                    + "ELSE 'exit'";
//bindings array
Object[] bindings = new Object[]{"REASON1","REASON2"};

dslContext.resultQuery(sql,  bindings); //this will fail with Caused by: org.postgresql.util.PSQLException: No value specified for parameter 2.
```

this query executes successfully on JOOQ 3.9.3 but fails in new versions

### Versions

- jOOQ: 3.14.7
- Java: 1.8
- Database (include vendor): PostgreSql 9.6
- OS: Windows 10
- JDBC Driver (include name if inofficial driver):postgresql-42.0.0.jar

**Comment by lukaseder** [comment]
This works perfectly fine for me. Can you please provide complete reproduction steps?

**Comment by kochetkovSergeY** [comment]
@lukaseder  sorry i edited the source sql, sql was with line breaks. This main class fails with error

import org.jooq.Record;
import org.jooq.Result;
import org.jooq.impl.DataSourceConnectionProvider;
import org.jooq.impl.DefaultConfiguration;
import org.jooq.impl.DefaultDSLContext;
import org.springframework.jdbc.datasource.DriverManagerDataSource;

public class JooqTest {

  public static void main(String[] args) {
    DriverManagerDataSource dataSource = new DriverManagerDataSource();
    dataSource.setUsername("user");
    dataSource.setPassword("pass");
    dataSource.setDriverClassName("org.postgresql.Driver");
    dataSource.setUrl("jdbc:postgresql://host:5432/db");
    DataSourceConnectionProvider connectionProvider = new DataSourceConnectionProvider(
        dataSource);
    org.jooq.Configuration configuration = new DefaultConfiguration().derive(connectionProvider);
    DefaultDSLContext dslContext = new DefaultDSLContext(configuration);

    Object[] sqlArgs = new Object[]{"ASD", "DSF"};
    Result<Record> records = dslContext.resultQuery("SELECT CASE\n"
                                                    + "WHEN  ? ='REASON1'  and ?<> 'FAILURE'  THEN 'STATUS2'\n"
                                                    + "ELSE 'exit'", sqlArgs).fetch();
  }
}

**Comment by lukaseder** [comment]
Oh wow, yes indeed, I can reproduce it now, thanks. That's quite the bug! Will fix ASAP

**Comment by lukaseder** [comment]
Investigating this now. Curious, we do have some unit tests with newlines in them. Seems like an obvious thing to have tested so far and not something that regresses so easily. Removing the newlines again from your example still reproduces the problem. The actual problem is your usage of `?<>` which resembles one of those esoteric operators. Try this:

```sql
select distinct oprname
from pg_operator
where oprname like '?%'
order by 1;
```

On my PG 13 installation, this yields:

```
?
?#
?&
?-
?-|
?|
?||
```

Our plain SQL templating "parser" implements a heuristic. Whenever `?` is followed by any of these characters: `?`, `|`, `&`, `@`, `<`, `~`, `#`, `-`, then we assume it's an operator. See also #5307, #7035, #7037.

That parser isn't really aware of SQL syntax, just of a few things like string literals, comments, and other things that change the semantics of the `?` marker.

Yours is quite the edge case. If you had added a space between `?` and `<>`, then the issue would not have appeared. Also, if you had used `!=`, then it would not have appeared either. Or if you had inversed the bind variable and the literal you're comparing it with, etc.

Not to say that this isn't very annoying. It should not have happened in your case. I guess we'll just risk it and exclude `?<>` explicitly from our operator heuristic... hoping we're not compromising again on potential future operators in PostgreSQL. (e.g. `?<@` is an `ltree` operator: https://www.postgresql.org/docs/current/ltree.html#id-1.11.7.30.6).

**Comment by lukaseder** [comment]
In any case, the workaround is to just add a space between `? <>`, like you did with `? =`

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.8 (#11444). Thanks again for your report.
